File: apiome-ui/examples/pydantic/07-composition-inheritance.py
"""Within-document composition: a mixin plus a base model, two specialisations, a
generic envelope instantiated by subclassing, and a model composed of other models."""

from __future__ import annotations

from datetime import datetime
from typing import Generic, Literal, Optional, TypeVar, Union

from pydantic import BaseModel, ConfigDict, Field

T = TypeVar("T")


class TimestampMixin(BaseModel):
    created_at: datetime
    updated_at: Optional[datetime] = None


class IdentifiedBase(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str = Field(pattern=r"^[A-Z]{3}-[0-9]{6}$")


class Record(IdentifiedBase, TimestampMixin):
    """Multiple inheritance: fields from both parents are merged."""

    label: str = Field(min_length=1, max_length=80)


class ActiveRecord(Record):
    status: Literal["active"] = "active"
    updated_at: datetime


class ArchivedRecord(Record):
    status: Literal["archived"] = "archived"
    archived_at: datetime


AnyRecord = Union[ActiveRecord, ArchivedRecord]


class Page(BaseModel, Generic[T]):
    items: list[T]
    total: int = Field(ge=0)
    next_cursor: Optional[str] = None


class RecordPage(Page[AnyRecord]):
    """Generic instantiated by subclassing — statically resolvable."""


class RecordSet(BaseModel):
    """Composed of other models rather than of scalars."""

    active: list[ActiveRecord] = Field(default_factory=list)
    archived: list[ArchivedRecord] = Field(default_factory=list)
    by_id: dict[str, Record] = Field(default_factory=dict)
    page: Optional[RecordPage] = None

File: apiome-ui/examples/typescript-types/07-composition-inheritance.ts
/**
 * Within-document composition: interface extension from two parents, intersection
 * aliases, a generic instantiated locally, and a union assembled from the results.
 */

export interface Timestamped {
  createdAt: string;
  updatedAt?: string;
}

export interface Identified {
  id: string;
}

/** Extension from two interfaces: members merge. */
export interface Record extends Identified, Timestamped {
  label: string;
}

export interface ActiveRecord extends Record {
  status: 'active';
  updatedAt: string;
}

export interface ArchivedRecord extends Record {
  status: 'archived';
  archivedAt: string;
}

export type AnyRecord = ActiveRecord | ArchivedRecord;

/** Intersection alias rather than interface extension — the other composition form. */
export type AuditedRecord = Record & {
  auditedBy: string;
  auditedAt: string;
};

export interface Page<T> {
  items: T[];
  total: number;
  nextCursor?: string;
}

/** Generic instantiated in this file, so it is resolvable. */
export type RecordPage = Page<AnyRecord>;

export interface RecordSet {
  active: ActiveRecord[];
  archived: ArchivedRecord[];
  byId: Record2Map;
  page?: RecordPage;
}

/** Index signature over a composed type. */
export interface Record2Map {
  [id: string]: AnyRecord;
}
